### PR TITLE
Speed up NEP-MD by tabulating g_n

### PR DESCRIPTION
**Pull request by brucefan1983**
In NEP, the $g_n$ radial functions are linear combinations of some basis functions:

$$
g_n(r_{ij}) = \sum_k c_{nk}^{ij} f_k(r_{ij})
$$

When the number of basis functions is large (which is required to achieve high accuracy), the computation of $g_n$ is quite expensive. Here I propose to use linear table to represent these functions, which can speed up the calculations to some degree while introducing little errors when the table is large enough. With the table, the computational cost in MD simulations is then independent of the basis size. So the speedup is higher when the basis size is larger.

The figure below shows the scalings of the speedup factor and the root-mean-square error (RMSE) with respect to the table length. I think a table lenght of 4000 is a very safe choice, while a table length of 1000 might also be good enough. The current code uses 2000 and one also needs to add `-DUSE_TABLE` in `makefile` to enable this feature. We can make it the default after doing extensive tests in MD simulations.

image

The figure below is for amorphous carbon with a table length of 4000:

image

The figure below is for amorphous carbon with a table length of 2500:

image

The figure below is for PbTe with a table length of 2000. The difference is close to single-precision level.
image

### CODE REVIEW

**Comment by brucefan1983** [comment]
sorry I used "edit" instead of "reply".

**Comment by brucefan1983** [comment]
This is very similar the the spline in the LAMMPS implementation of the EAM potential, but I have chosen the simpler linear interpolation. For cubic spline, one needs to read in a few values for each distance.

**Comment by elindgren** [comment]
> For each , I pre-calculate a table of 2001 function values at . Then in the force calcualion kernels, for each and distance , only need to read in two values in the table, say
and
and use linear interpolation to get an approximate value of .

I see, neat!

File: src/utilities/nep_utilities.cuh [hunk 1]
```diff
@@ -776,3 +776,76 @@ find_q_with_5body(const int n_max_angular_plus_1, const int n, const float* s, f
...
+  float& weight_right)
+{
+  float d12_index = d12_reduced * table_segments;
+  index_left = int(d12_index);
+  if (index_left == table_segments) {
+    --index_left;
+  }
+  index_right = index_left + 1;
+  weight_right = d12_index - index_left;
+  weight_left = 1.0f - weight_right;
```
Comment (elindgren) [inline src/utilities/nep_utilities.cuh:801]: From this I understand that `weight_right` is indeed the "fraction" of the right table value, and equivalently for the left. Neat!

**Review by elindgren — COMMENTED** [review]
Looks good to me, I had a few questions though.

File: src/force/nep3.cu [hunk 2]
```diff
@@ -436,6 +479,21 @@ static __global__ void find_descriptor(
       apply_mic(box, x12double, y12double, z12double);
       float x12 = float(x12double), y12 = float(y12double), z12 = float(z12double);
       float d12 = sqrt(x12 * x12 + y12 * y12 + z12 * z12);
+
+#ifdef USE_TABLE
+      int index_left, index_right;
+      float weight_left, weight_right;
```
[the remark this answers is not in the document]
Comment (brucefan1983) [reply src/force/nep3.cu:485]: yes

File: src/force/nep3.cu [hunk 3]
```diff
@@ -691,10 +797,37 @@ static __global__ void find_partial_force_angular(
...
+      find_index_and_weight(
+        d12 * paramb.rcinv_angular, index_left, index_right, weight_left, weight_right);
+      int t12 = t1 * paramb.num_types + g_type[n2];
+      for (int n = 0; n <= paramb.n_max_angular; ++n) {
+        int index_left_all =
+          (index_left * paramb.num_types_sq + t12) * (paramb.n_max_angular + 1) + n;
+        int index_right_all =
+          (index_right * paramb.num_types_sq + t12) * (paramb.n_max_angular + 1) + n;
```
[the remark this answers is not in the document]
Comment (brucefan1983) [reply src/force/nep3.cu:811]: the table is logically a 3D array, like `gn[table_segments][num_teyps_sq][n_max+1]`, so an element `gn[index_left][t12][n]` is indexed such as
```
gn[(index_left * num_types_sq + t12) * (n_max + 1) + n];
```
using 1D notation.

File: src/utilities/nep_utilities.cuh [hunk 4]
```diff
@@ -776,3 +776,76 @@ find_q_with_5body(const int n_max_angular_plus_1, const int n, const float* s, f
   q[5 * n_max_angular_plus_1 + n] = C5B[0] * s0_sq * s0_sq + C5B[1] * s0_sq * s1_sq_plus_s2_sq +
                                     C5B[2] * s1_sq_plus_s2_sq * s1_sq_plus_s2_sq;
 }
+
+#ifdef USE_TABLE
+namespace
+{
+const int table_length = 2001;
```
[the remark this answers is not in the document]
Comment (brucefan1983) [reply src/utilities/nep_utilities.cuh:783]: using keyword is also doable, but I started from the simplest approach: using a conditional compilation. I will only bother to do it only if this turns out to be a very promising approach.

File: src/utilities/nep_utilities.cuh (hunk 1 above)
Comment (brucefan1983) [reply src/utilities/nep_utilities.cuh:801]: Rright!

**Review by hityingph — APPROVED** [review]
I tested two systems, namely monolayer MoSe2 and WSe2, using unpublished nep models. I discovered that using the linear table approximation resulted in a 10% increase in speed and allowed me to obtain consistent thermal conductivity values when compared to the pristine nep models using HNEMD simulations.

**Comment by brucefan1983** [comment]
> 

thanks for testing!

File: src/force/nep3.cu (hunk 3 above)
[the remark this answers is not in the document]
Comment (elindgren) [reply src/force/nep3.cu:811]: Ah right :+1:

File: src/utilities/nep_utilities.cuh (hunk 4 above)
[the remark this answers is not in the document]
Comment (elindgren) [reply src/utilities/nep_utilities.cuh:783]: Makes sense!

**Comment by shdchen** [comment]
We recently tested some multiple-element NEPs and got quite remarkable speedup. Some speedup factors  turned out to be more than 30%. Very impressive.

**Comment by brucefan1983** [comment]
Sure, I can do it soon.

William ***@***.***> 于 2023年3月29日周三 17:28写道：

**Comment by ZKC19940412** [comment]
A relevant question along the line: Is this linear table interpolation works for order vision of NEP potential (say somewhere between NEP 3.3 to 3.4)? Or that one has to retrain the potential (even just one epoch) using the newer vision of NEP?

**Comment by brucefan1983** [comment]
The speedup  works for all versions of NEP, if the`nep.txt` file can be correctly read in by the code. Only need to add `-DUSE_TABLE` to `CFLAGS` in makefile. However, NEP2 models will probably not see an actual speedup as there are no basis functions in NEP2.

If you got some errors for reading the `nep.txt` file, it means your NEP model was trained during some format-trainsition stage and you can contact me. It just needs some manual changes to the first few lines in `nep.txt`.

**Comment by brucefan1983** [comment]
> Are there plans to upstream this to the [NEP_CPU](https://github.com/brucefan1983/NEP_CPU)?

I have made a PR In the NEP_CPU repo: 
https://github.com/brucefan1983/NEP_CPU/pull/18

Could you test it with LAMMPS to see if there is expected speedup and if all the results are reasonable? I have only tested the non-LAMMPS part using my local tests.

**Pull request merged**